cli/exp_task_status.go

```diff
@@ -0,0 +1,152 @@
+package cli
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/cli/cliui"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/serpent"
+)
+
+func (r *RootCmd) taskStatus() *serpent.Command {
+	var (
+		client    = new(codersdk.Client)
+		formatter = cliui.NewOutputFormatter(
+			cliui.TableFormat(
+				[]taskStatusRow{},
+				[]string{
+					"state changed",
+					"status",
+					"state",
+					"message",
+				},
+			),
+			cliui.JSONFormat(),
+		)
+		watchArg         bool
+		watchIntervalArg time.Duration
+	)
+	cmd := &serpent.Command{
+		Short:   "Show the status of a task.",
+		Use:     "status",
+		Aliases: []string{"stat"},
+		Options: serpent.OptionSet{
+			{
+				Default:     "false",
+				Description: "Watch the task status output. This will stream updates to the terminal until the underlying workspace is stopped.",
+				Flag:        "watch",
+				Name:        "watch",
+				Value:       serpent.BoolOf(&watchArg),
+			},
+			{
+				Default:     "1s",
+				Description: "Interval to poll the task for updates. Only used in tests.",
+				Hidden:      true,
+				Flag:        "watch-interval",
+				Name:        "watch-interval",
+				Value:       serpent.DurationOf(&watchIntervalArg),
+			},
+		},
+		Middleware: serpent.Chain(
+			serpent.RequireNArgs(1),
+			r.InitClient(client),
+		),
+		Handler: func(i *serpent.Invocation) error {
+			ctx := i.Context()
+			ec := codersdk.NewExperimentalClient(client)
+			id := i.Args[0]
+
+			// TODO: right now tasks are still "workspaces" under the hood.
+			// We should update this once we have a proper task model.
+			ws, err := namedWorkspace(ctx, client, id)
+			if err != nil {
+				return err
+			}
+
+			task, err := ec.TaskByID(ctx, ws.ID)
+			if err != nil {
+				return err
+			}
+
+			out, err := formatter.Format(ctx, toStatusRow(task))
+			if err != nil {
+				return xerrors.Errorf("format task status: %w", err)
+			}
+			_, _ = fmt.Fprintln(i.Stdout, out)
+
+			if !watchArg {
+				return nil
+			}
+
+			lastStatus := task.Status
+			lastState := task.CurrentState
+			t := time.NewTicker(watchIntervalArg)
+			defer t.Stop()
+			// TODO: implement streaming updates instead of polling
+			for range t.C {
+				task, err := ec.TaskByID(ctx, ws.ID)
+				if err != nil {
+					return err
+				}
+				if lastStatus == task.Status && taskStatusEqual(lastState, task.CurrentState) {
+					continue
+				}
+				out, err := formatter.Format(ctx, toStatusRow(task))
+				if err != nil {
+					return xerrors.Errorf("format task status: %w", err)
+				}
+				// hack: skip the extra column header from formatter
+				if formatter.FormatID() != cliui.JSONFormat().ID() {
+					out = strings.SplitN(out, "\n", 2)[1]
+				}
+				_, _ = fmt.Fprintln(i.Stdout, out)
```

Comment:
With output json, this seems to output arrays, wdyt about making it plain object? Seems more natural. Outputting in a single line might also make it easier to use but dealers choice.

If you look at `exp task list` there's a way to output in a different format, we could potentially use the raw task object for json output since it has all the info the user may want and is a bit more familiar than a bespoke object?

Reply:
Updated to use raw task for JSON output.